api/agent/agent.go

```diff
@@ -113,13 +113,39 @@ type agent struct {
 	// TODO(reed): shoot this fucking thing
 	callOverrider CallOverrider
 
+	// Call Extension interceptor
+	extensionInterceptor ExtensionInterceptor
+
 	// additional options to configure each call
 	callOpts []CallOpt
 
 	// deferred actions to call at end of initialisation
 	onStartup []func()
 }
 
+type ExtensionInterceptor interface {
+	// BeforeCreate is triggered immediately preceding container creation, after all other
+	// configuration is in place.
+	// ContainerTask.Id() can be used to provide a stable per-container key.
+	BeforeCreate(context.Context, drivers.ContainerTask, drivers.Cookie, Call) error
+
+	// AfterDestroy is triggered immediately after container shutdown.
+	// ContainerTask.Id() can be used to provide a stable per-container key.
+	AfterDestroy(context.Context, drivers.ContainerTask) error
```

Comment:
at this point we have already returned data back to LB? is that the reason because we don't have the Call object here?

Reply:
Yeah, this is afterwards, on eviction or other forms of container shutdown.